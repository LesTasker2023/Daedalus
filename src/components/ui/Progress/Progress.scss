// Progress component SCSS enhancements
// Use this file for ENHANCEMENTS that can't be achieved with Tailwind classes
// Avoid duplicating what Tailwind already provides

.progress {
  // Smooth transitions for all interactive states (STANDARDIZED)
  transition: all 0.15s ease-in-out;

  // Striped pattern
  &--striped {
    .progress__indicator {
      background-image: linear-gradient(
        45deg,
        rgba(255, 255, 255, 0.15) 25%,
        transparent 25%,
        transparent 50%,
        rgba(255, 255, 255, 0.15) 50%,
        rgba(255, 255, 255, 0.15) 75%,
        transparent 75%,
        transparent
      );
      background-size: 1rem 1rem;
    }
  }

  // Animated stripes
  &--animated {
    .progress__indicator--animated {
      animation: progress-stripes 1s linear infinite;
    }
  }

  // Progress wrapper for outside labels
  &-wrapper {
    display: flex;
    align-items: center;
    gap: 0.5rem;
    width: 100%;
  }

  // Label styling
  &__label {
    font-size: 0.75rem;
    font-weight: 500;
    white-space: nowrap;

    &--inside {
      display: flex;
      align-items: center;
      justify-content: center;
      height: 100%;
      color: white;
      text-shadow: 0 1px 2px rgba(0, 0, 0, 0.5);
      min-width: 2rem;
    }

    &--outside {
      color: hsl(var(--foreground));
      min-width: 2.5rem;
      text-align: right;
    }
  }

  // Error state styling (STANDARDIZED ADDITION)
  &--error {
    .progress__indicator {
      background-color: hsl(var(--destructive));
    }

    // Error state with stripes
    &.progress--striped .progress__indicator {
      background-image: linear-gradient(
        45deg,
        rgba(255, 255, 255, 0.15) 25%,
        transparent 25%,
        transparent 50%,
        rgba(255, 255, 255, 0.15) 50%,
        rgba(255, 255, 255, 0.15) 75%,
        transparent 75%,
        transparent
      );
    }
  }
}

// Keyframe animations
@keyframes progress-stripes {
  0% {
    background-position: 1rem 0;
  }
  100% {
    background-position: 0 0;
  }
}

// Pulse animation for indeterminate progress
@keyframes progress-pulse {
  0%,
  100% {
    opacity: 1;
  }
  50% {
    opacity: 0.5;
  }
}

// Enhanced focus styles for accessibility
.progress {
  &:focus-visible {
    outline: 2px solid hsl(var(--ring));
    outline-offset: 2px;
  }
}

// High contrast support
@media (prefers-contrast: high) {
  .progress {
    border: 1px solid hsl(var(--border));

    &__indicator {
      border-right: 2px solid hsl(var(--background));
    }
  }
}

// Reduced motion support
@media (prefers-reduced-motion: reduce) {
  .progress {
    &__indicator {
      transition: none !important;
    }

    &--animated {
      .progress__indicator--animated {
        animation: none !important;
      }
    }
  }
}
