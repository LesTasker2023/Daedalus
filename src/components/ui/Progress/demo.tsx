'use client';

import React, { useState } from 'react';
import { Progress } from '.';

export default function ProgressDemo() {
  const [state, setState] = useState(false);

  return (
    <div className="space-y-6 p-6 max-w-2xl mx-auto">
      <div className="space-y-2">
        <h1 className="text-2xl font-bold">Progress Component</h1>
        <p className="text-muted-foreground">Progress indicators</p>
      </div>

      <div className="space-y-4">
        <div className="space-y-2">
          <h3 className="text-lg font-semibold">Basic Usage</h3>
          <div className="border rounded-lg p-4 bg-background">
            <Progress>Progress Example</Progress>
          </div>
        </div>

        <div className="space-y-2">
          <h3 className="text-lg font-semibold">Interactive</h3>
          <div className="border rounded-lg p-4 bg-background space-y-2">
            <Progress onClick={() => setState(!state)}>
              Click me {state ? '(Active)' : '(Inactive)'}
            </Progress>
            <p className="text-sm text-muted-foreground">State: {state ? 'Active' : 'Inactive'}</p>
          </div>
        </div>
      </div>
    </div>
  );
}
