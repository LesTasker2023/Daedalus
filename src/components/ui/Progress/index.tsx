export { default as ProgressDemo } from './demo';
export * from './progress';
export type { ProgressProps } from './progress';
