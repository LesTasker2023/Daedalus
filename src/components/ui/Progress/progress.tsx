'use client';

import * as ProgressPrimitive from '@radix-ui/react-progress';
import { cva, type VariantProps } from 'class-variance-authority';
import * as React from 'react';

import { cn } from '@/lib/utils';

const progressVariants = cva('relative w-full overflow-hidden rounded-full bg-primary/20', {
  variants: {
    size: {
      sm: 'h-1',
      default: 'h-2',
      lg: 'h-3',
      xl: 'h-4',
    },
    variant: {
      default: 'bg-primary/20',
      success: 'bg-green-100 dark:bg-green-900/20',
      warning: 'bg-yellow-100 dark:bg-yellow-900/20',
      danger: 'bg-red-100 dark:bg-red-900/20',
    },
  },
  defaultVariants: {
    size: 'default',
    variant: 'default',
  },
});

const indicatorVariants = cva('h-full w-full flex-1 transition-all duration-300 ease-in-out', {
  variants: {
    variant: {
      default: 'bg-primary',
      success: 'bg-green-500',
      warning: 'bg-yellow-500',
      danger: 'bg-red-500',
    },
  },
  defaultVariants: {
    variant: 'default',
  },
});

export interface ProgressProps
  extends React.ComponentPropsWithoutRef<typeof ProgressPrimitive.Root>,
    VariantProps<typeof progressVariants> {}

const Progress = React.forwardRef<React.ElementRef<typeof ProgressPrimitive.Root>, ProgressProps>(
  ({ className, value, size, variant, ...props }, ref) => (
    <ProgressPrimitive.Root
      ref={ref}
      className={cn(progressVariants({ size, variant }), className)}
      value={value}
      {...props}
    >
      <ProgressPrimitive.Indicator
        className={cn(indicatorVariants({ variant }), 'rounded-full')}
        style={{ transform: `translateX(-${100 - (value || 0)}%)` }}
      />
    </ProgressPrimitive.Root>
  )
);
Progress.displayName = ProgressPrimitive.Root.displayName;

export { Progress };
