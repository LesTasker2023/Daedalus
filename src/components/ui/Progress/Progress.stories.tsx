import type { Meta, StoryObj } from '@storybook/nextjs';
import Progress from './progress';

const meta = {
  title: 'UI/Progress',
  component: Progress,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component: 'Enhanced Progress component with variants, sizes, labels, and animations.',
      },
    },
  },
  argTypes: {
    value: {
      control: { type: 'range', min: 0, max: 100, step: 1 },
    },
    variant: {
      control: 'select',
      options: ['default', 'success', 'warning', 'danger'],
    },
    size: {
      control: 'select',
      options: ['sm', 'default', 'lg', 'xl'],
    },
    showLabel: {
      control: 'boolean',
    },
    labelPosition: {
      control: 'select',
      options: ['inside', 'outside'],
    },
    animated: {
      control: 'boolean',
    },
    striped: {
      control: 'boolean',
    },
  },
  tags: ['autodocs'],
} satisfies Meta<typeof Progress>;

export default meta;
type Story = StoryObj<typeof meta>;

// 1. All Variants
export const AllVariants: Story = {
  render: () => (
    <div className="space-y-4 w-80">
      <div>
        <h4 className="text-sm font-medium mb-2">Default</h4>
        <Progress value={65} variant="default" />
      </div>
      <div>
        <h4 className="text-sm font-medium mb-2">Success</h4>
        <Progress value={85} variant="success" />
      </div>
      <div>
        <h4 className="text-sm font-medium mb-2">Warning</h4>
        <Progress value={45} variant="warning" />
      </div>
      <div>
        <h4 className="text-sm font-medium mb-2">Danger</h4>
        <Progress value={25} variant="danger" />
      </div>
    </div>
  ),
};

// 2. All Sizes
export const AllSizes: Story = {
  render: () => (
    <div className="space-y-4 w-80">
      <div>
        <h4 className="text-sm font-medium mb-2">Small</h4>
        <Progress value={75} size="sm" />
      </div>
      <div>
        <h4 className="text-sm font-medium mb-2">Default</h4>
        <Progress value={75} size="default" />
      </div>
      <div>
        <h4 className="text-sm font-medium mb-2">Large</h4>
        <Progress value={75} size="lg" />
      </div>
      <div>
        <h4 className="text-sm font-medium mb-2">Extra Large</h4>
        <Progress value={75} size="xl" />
      </div>
    </div>
  ),
};

// 3. Interactive Example
export const InteractiveExample: Story = {
  args: {
    value: 60,
    variant: 'default',
    size: 'default',
    showLabel: false,
    labelPosition: 'outside',
    animated: false,
    striped: false,
  },
  render: args => (
    <div className="w-80">
      <Progress {...args} />
    </div>
  ),
};

// 4. Enhanced Features
export const EnhancedFeatures: Story = {
  render: () => (
    <div className="space-y-6 w-80">
      <div>
        <h4 className="text-sm font-medium mb-2">With Outside Labels</h4>
        <div className="space-y-2">
          <Progress value={25} variant="danger" showLabel labelPosition="outside" />
          <Progress value={60} variant="warning" showLabel labelPosition="outside" />
          <Progress value={85} variant="success" showLabel labelPosition="outside" />
        </div>
      </div>

      <div>
        <h4 className="text-sm font-medium mb-2">With Inside Labels</h4>
        <div className="space-y-2">
          <Progress value={45} variant="default" size="lg" showLabel labelPosition="inside" />
          <Progress value={70} variant="success" size="lg" showLabel labelPosition="inside" />
          <Progress value={90} variant="warning" size="lg" showLabel labelPosition="inside" />
        </div>
      </div>

      <div>
        <h4 className="text-sm font-medium mb-2">Striped Progress</h4>
        <div className="space-y-2">
          <Progress value={40} variant="default" striped />
          <Progress value={65} variant="success" striped />
          <Progress value={85} variant="warning" striped />
        </div>
      </div>

      <div>
        <h4 className="text-sm font-medium mb-2">Animated Stripes</h4>
        <div className="space-y-2">
          <Progress value={50} variant="default" striped animated />
          <Progress value={75} variant="success" striped animated />
        </div>
      </div>
    </div>
  ),
};

// 5. Real-world Usage
export const ProgressStates: Story = {
  render: () => (
    <div className="space-y-6 w-80">
      <div>
        <h4 className="text-sm font-medium mb-2">File Upload Progress</h4>
        <div className="space-y-3">
          <div>
            <div className="flex justify-between text-sm mb-1">
              <span>document.pdf</span>
              <span>45%</span>
            </div>
            <Progress value={45} variant="default" />
          </div>
          <div>
            <div className="flex justify-between text-sm mb-1">
              <span>image.jpg</span>
              <span>Complete</span>
            </div>
            <Progress value={100} variant="success" />
          </div>
        </div>
      </div>

      <div>
        <h4 className="text-sm font-medium mb-2">System Resources</h4>
        <div className="space-y-3">
          <div>
            <div className="flex justify-between text-sm mb-1">
              <span>CPU Usage</span>
              <span>72%</span>
            </div>
            <Progress value={72} variant="warning" size="sm" />
          </div>
          <div>
            <div className="flex justify-between text-sm mb-1">
              <span>Memory</span>
              <span>89%</span>
            </div>
            <Progress value={89} variant="danger" size="sm" />
          </div>
          <div>
            <div className="flex justify-between text-sm mb-1">
              <span>Storage</span>
              <span>34%</span>
            </div>
            <Progress value={34} variant="success" size="sm" />
          </div>
        </div>
      </div>
    </div>
  ),
};

// 6. Progress Steps
export const ProgressSteps: Story = {
  render: () => (
    <div className="space-y-6 w-80">
      <div>
        <h4 className="text-sm font-medium mb-2">Multi-Step Process</h4>
        <div className="space-y-3">
          <div>
            <div className="text-sm mb-1">Step 1: Account Setup</div>
            <Progress value={100} variant="success" size="sm" />
          </div>
          <div>
            <div className="text-sm mb-1">Step 2: Profile Information</div>
            <Progress value={100} variant="success" size="sm" />
          </div>
          <div>
            <div className="text-sm mb-1">Step 3: Payment Details</div>
            <Progress value={60} variant="default" size="sm" />
          </div>
          <div>
            <div className="text-sm mb-1 text-muted-foreground">Step 4: Confirmation</div>
            <Progress value={0} variant="default" size="sm" />
          </div>
        </div>
      </div>
    </div>
  ),
};

// 7. Size and Variant Matrix
export const SizeVariantMatrix: Story = {
  render: () => (
    <div className="space-y-6 w-96">
      {(['sm', 'default', 'lg', 'xl'] as const).map(size => (
        <div key={size}>
          <h4 className="text-sm font-medium mb-2 capitalize">{size} Size</h4>
          <div className="space-y-2">
            <Progress value={65} variant="default" size={size} />
            <Progress value={75} variant="success" size={size} />
            <Progress value={45} variant="warning" size={size} />
            <Progress value={25} variant="danger" size={size} />
          </div>
        </div>
      ))}
    </div>
  ),
};

// 8. Loading States
export const LoadingStates: Story = {
  render: () => (
    <div className="space-y-6 w-80">
      <div>
        <h4 className="text-sm font-medium mb-2">Loading Animation</h4>
        <Progress value={45} variant="default" striped animated showLabel labelPosition="outside" />
      </div>

      <div>
        <h4 className="text-sm font-medium mb-2">Success State</h4>
        <Progress value={100} variant="success" showLabel labelPosition="outside" />
      </div>

      <div>
        <h4 className="text-sm font-medium mb-2">Error State</h4>
        <Progress value={30} variant="danger" showLabel labelPosition="outside" />
      </div>
    </div>
  ),
};

// 9. Accessibility States
export const AccessibilityStates: Story = {
  render: () => (
    <div className="space-y-6 w-80">
      <div>
        <h4 className="text-sm font-medium mb-2">With ARIA Labels</h4>
        <Progress
          value={65}
          variant="default"
          showLabel
          labelPosition="outside"
          aria-label="Download progress"
          aria-describedby="progress-description"
        />
        <p id="progress-description" className="text-xs text-muted-foreground mt-1">
          Downloading file: 65% complete
        </p>
      </div>

      <div>
        <h4 className="text-sm font-medium mb-2">High Contrast Mode</h4>
        <div className="space-y-2 p-3 border rounded" style={{ filter: 'contrast(150%)' }}>
          <Progress value={40} variant="default" />
          <Progress value={70} variant="success" />
          <Progress value={50} variant="warning" />
        </div>
      </div>
    </div>
  ),
};

// 10. Edge Cases
export const EdgeCases: Story = {
  render: () => (
    <div className="space-y-6 w-80">
      <div>
        <h4 className="text-sm font-medium mb-2">Boundary Values</h4>
        <div className="space-y-2">
          <div className="text-xs text-muted-foreground">0% Progress</div>
          <Progress value={0} variant="default" showLabel labelPosition="outside" />

          <div className="text-xs text-muted-foreground">100% Progress</div>
          <Progress value={100} variant="success" showLabel labelPosition="outside" />

          <div className="text-xs text-muted-foreground">Undefined Value</div>
          <Progress variant="default" showLabel labelPosition="outside" />
        </div>
      </div>

      <div>
        <h4 className="text-sm font-medium mb-2">Extreme Sizes</h4>
        <div className="space-y-2">
          <div className="text-xs text-muted-foreground">Very Narrow</div>
          <div className="w-20">
            <Progress value={75} variant="default" showLabel labelPosition="outside" />
          </div>

          <div className="text-xs text-muted-foreground">Very Wide</div>
          <div className="w-full max-w-2xl">
            <Progress value={45} variant="warning" showLabel labelPosition="inside" size="lg" />
          </div>
        </div>
      </div>
    </div>
  ),
};
