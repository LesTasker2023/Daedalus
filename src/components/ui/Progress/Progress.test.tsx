import { render, screen } from '@testing-library/react';
import React from 'react';
import { describe, expect, it } from 'vitest';
import Progress from './progress';

describe('Progress Component', () => {
  // 1. Rendering Tests (7 tests)
  describe('Rendering', () => {
    it('renders progress with value', () => {
      render(<Progress value={50} />);
      const progress = screen.getByRole('progressbar');
      expect(progress).toBeInTheDocument();
    });

    it('renders with default value of 0 when no value provided', () => {
      render(<Progress />);
      const progress = screen.getByRole('progressbar');
      expect(progress).toBeInTheDocument();
    });

    it('forwards ref correctly', () => {
      const ref = React.createRef<HTMLDivElement>();
      render(<Progress ref={ref} value={50} />);
      expect(ref.current).toBeInstanceOf(HTMLDivElement);
    });

    it('applies custom className', () => {
      render(<Progress value={50} className="custom-class" />);
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveClass('custom-class');
    });

    it('passes through HTML attributes', () => {
      render(<Progress value={50} data-testid="progress-test" aria-label="Test progress" />);
      const progress = screen.getByTestId('progress-test');
      expect(progress).toHaveAttribute('aria-label', 'Test progress');
    });

    it('renders with correct progress indicator transform', () => {
      render(<Progress value={75} />);
      const progress = screen.getByRole('progressbar');
      const indicator = progress.firstElementChild as HTMLElement;
      expect(indicator).toHaveStyle('transform: translateX(-25%)');
    });

    it('displays correct displayName', () => {
      expect(Progress.displayName).toBe('Progress');
    });
  });

  // 2. Variants Tests (4 tests)
  describe('Variants', () => {
    it('renders default variant correctly', () => {
      render(<Progress value={50} variant="default" />);
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveClass('bg-primary/20');
    });

    it('renders success variant correctly', () => {
      render(<Progress value={50} variant="success" />);
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveClass('bg-green-100', 'dark:bg-green-900/20');
    });

    it('renders warning variant correctly', () => {
      render(<Progress value={50} variant="warning" />);
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveClass('bg-yellow-100', 'dark:bg-yellow-900/20');
    });

    it('renders danger variant correctly', () => {
      render(<Progress value={50} variant="danger" />);
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveClass('bg-red-100', 'dark:bg-red-900/20');
    });
  });

  // 3. Sizes Tests (4 tests)
  describe('Sizes', () => {
    it('renders small size correctly', () => {
      render(<Progress value={50} size="sm" />);
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveClass('h-1');
    });

    it('renders default size correctly', () => {
      render(<Progress value={50} size="default" />);
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveClass('h-2');
    });

    it('renders large size correctly', () => {
      render(<Progress value={50} size="lg" />);
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveClass('h-3');
    });

    it('renders extra large size correctly', () => {
      render(<Progress value={50} size="xl" />);
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveClass('h-4');
    });
  });

  // 4. Enhanced Features Tests (12 tests)
  describe('Enhanced Features', () => {
    describe('Label functionality', () => {
      it('does not show label by default', () => {
        render(<Progress value={75} />);
        expect(screen.queryByText('75%')).not.toBeInTheDocument();
      });

      it('shows outside label when showLabel is true', () => {
        render(<Progress value={75} showLabel />);
        expect(screen.getByText('75%')).toBeInTheDocument();
        expect(screen.getByText('75%')).toHaveClass('progress__label--outside');
      });

      it('shows inside label when labelPosition is inside', () => {
        render(<Progress value={75} showLabel labelPosition="inside" />);
        expect(screen.getByText('75%')).toBeInTheDocument();
        expect(screen.getByText('75%')).toHaveClass('progress__label--inside');
      });

      it('renders wrapper div when outside label is shown', () => {
        render(<Progress value={75} showLabel labelPosition="outside" />);
        const wrapper = screen.getByText('75%').closest('.progress-wrapper');
        expect(wrapper).toBeInTheDocument();
      });

      it('does not render wrapper div when inside label is shown', () => {
        render(<Progress value={75} showLabel labelPosition="inside" />);
        const wrapper = document.querySelector('.progress-wrapper');
        expect(wrapper).not.toBeInTheDocument();
      });

      it('rounds percentage values correctly', () => {
        render(<Progress value={66.666} showLabel />);
        expect(screen.getByText('67%')).toBeInTheDocument();
      });
    });

    describe('Animation and styling features', () => {
      it('applies animated class when animated prop is true', () => {
        render(<Progress value={50} animated />);
        const progress = screen.getByRole('progressbar');
        expect(progress).toHaveClass('progress--animated');
      });

      it('applies striped class when striped prop is true', () => {
        render(<Progress value={50} striped />);
        const progress = screen.getByRole('progressbar');
        expect(progress).toHaveClass('progress--striped');
      });

      it('applies both animated and striped classes together', () => {
        render(<Progress value={50} animated striped />);
        const progress = screen.getByRole('progressbar');
        expect(progress).toHaveClass('progress--animated', 'progress--striped');
      });

      it('applies indicator animation class when animated', () => {
        render(<Progress value={50} animated />);
        const progress = screen.getByRole('progressbar');
        const indicator = progress.firstElementChild as HTMLElement;
        expect(indicator).toHaveClass('progress__indicator--animated');
      });

      it('applies correct indicator variant classes', () => {
        render(<Progress value={50} variant="success" />);
        const progress = screen.getByRole('progressbar');
        const indicator = progress.firstElementChild as HTMLElement;
        expect(indicator).toHaveClass('bg-green-500');
      });

      it('applies correct transition duration', () => {
        render(<Progress value={50} />);
        const progress = screen.getByRole('progressbar');
        const indicator = progress.firstElementChild as HTMLElement;
        expect(indicator).toHaveClass('duration-300', 'ease-in-out');
      });
    });
  });

  // 5. Value Handling Tests (8 tests)
  describe('Value Handling', () => {
    it('handles zero value correctly', () => {
      render(<Progress value={0} />);
      const progress = screen.getByRole('progressbar');
      const indicator = progress.firstElementChild as HTMLElement;
      expect(indicator).toHaveStyle('transform: translateX(-100%)');
    });

    it('handles 100% value correctly', () => {
      render(<Progress value={100} />);
      const progress = screen.getByRole('progressbar');
      const indicator = progress.firstElementChild as HTMLElement;
      expect(indicator).toHaveStyle('transform: translateX(-0%)');
    });

    it('handles mid-range values correctly', () => {
      render(<Progress value={45} />);
      const progress = screen.getByRole('progressbar');
      const indicator = progress.firstElementChild as HTMLElement;
      expect(indicator).toHaveStyle('transform: translateX(-55%)');
    });

    it('handles undefined value as 0', () => {
      render(<Progress />);
      const progress = screen.getByRole('progressbar');
      const indicator = progress.firstElementChild as HTMLElement;
      expect(indicator).toHaveStyle('transform: translateX(-100%)');
    });

    it('handles decimal values correctly', () => {
      render(<Progress value={33.33} showLabel />);
      expect(screen.getByText('33%')).toBeInTheDocument();
      const progress = screen.getByRole('progressbar');
      const indicator = progress.firstElementChild as HTMLElement;
      expect(indicator).toHaveStyle('transform: translateX(-66.67%)');
    });

    it('handles negative values gracefully', () => {
      render(<Progress value={-10} showLabel />);
      const progress = screen.getByRole('progressbar');
      const indicator = progress.firstElementChild as HTMLElement;
      expect(indicator).toHaveStyle('transform: translateX(-110%)');
      expect(screen.getByText('-10%')).toBeInTheDocument();
    });

    it('handles values over 100 gracefully', () => {
      render(<Progress value={150} showLabel />);
      const progress = screen.getByRole('progressbar');
      const indicator = progress.firstElementChild as HTMLElement;
      expect(indicator).toHaveStyle('transform: translateX(--50%)');
      expect(screen.getByText('150%')).toBeInTheDocument();
    });

    it('handles very small decimal values', () => {
      render(<Progress value={0.1} showLabel />);
      expect(screen.getByText('0%')).toBeInTheDocument();
    });
  });

  // 6. Accessibility Tests (6 tests)
  describe('Accessibility', () => {
    it('has correct role attribute', () => {
      render(<Progress value={50} />);
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveAttribute('role', 'progressbar');
    });

    it('has correct default aria-valuemin and aria-valuemax', () => {
      render(<Progress value={50} />);
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveAttribute('aria-valuemin', '0');
      expect(progress).toHaveAttribute('aria-valuemax', '100');
    });

    it('supports custom aria attributes', () => {
      render(<Progress value={50} aria-label="File upload progress" />);
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveAttribute('aria-label', 'File upload progress');
    });

    it('supports aria-describedby for detailed descriptions', () => {
      render(
        <div>
          <Progress value={60} aria-describedby="progress-desc" />
          <div id="progress-desc">Uploading document.pdf</div>
        </div>
      );
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveAttribute('aria-describedby', 'progress-desc');
    });

    it('label text is accessible to screen readers', () => {
      render(<Progress value={85} showLabel labelPosition="outside" />);
      const label = screen.getByText('85%');
      expect(label).toBeInTheDocument();
    });

    it('maintains focus accessibility', () => {
      render(<Progress value={50} tabIndex={0} />);
      const progress = screen.getByRole('progressbar');
      progress.focus();
      expect(progress).toHaveFocus();
    });
  });

  // 7. Edge Cases Tests (6 tests)
  describe('Edge Cases', () => {
    it('handles missing variant prop gracefully', () => {
      render(<Progress value={50} />);
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveClass('bg-primary/20'); // default variant
    });

    it('handles missing size prop gracefully', () => {
      render(<Progress value={50} />);
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveClass('h-2'); // default size
    });

    it('handles combination of all enhanced features', () => {
      render(
        <Progress
          value={65}
          variant="success"
          size="lg"
          showLabel
          labelPosition="inside"
          animated
          striped
        />
      );
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveClass(
        'bg-green-100',
        'h-3',
        'progress--animated',
        'progress--striped'
      );
      expect(screen.getByText('65%')).toHaveClass('progress__label--inside');
    });

    it('handles rapid value changes', () => {
      const { rerender } = render(<Progress value={0} />);

      rerender(<Progress value={50} />);
      let progress = screen.getByRole('progressbar');
      let indicator = progress.firstElementChild as HTMLElement;
      expect(indicator).toHaveStyle('transform: translateX(-50%)');

      rerender(<Progress value={100} />);
      progress = screen.getByRole('progressbar');
      indicator = progress.firstElementChild as HTMLElement;
      expect(indicator).toHaveStyle('transform: translateX(-0%)');
    });

    it('handles very long custom classNames', () => {
      const longClassName =
        'very-long-custom-class-name-that-might-cause-issues-with-rendering-or-performance';
      render(<Progress value={50} className={longClassName} />);
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveClass(longClassName);
    });

    it('handles boolean props correctly', () => {
      render(<Progress value={50} animated={false} striped={false} showLabel={false} />);
      const progress = screen.getByRole('progressbar');
      expect(progress).not.toHaveClass('progress--animated', 'progress--striped');
      expect(screen.queryByText('50%')).not.toBeInTheDocument();
    });
  });
});
