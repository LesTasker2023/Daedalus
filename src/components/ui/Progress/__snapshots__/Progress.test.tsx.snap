// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Progress > Snapshots > matches default snapshot 1`] = `
<div
  aria-valuemax="100"
  aria-valuemin="0"
  class="relative w-full overflow-hidden rounded-full h-2 bg-primary/20"
  data-max="100"
  data-state="indeterminate"
  data-testid="progress"
  role="progressbar"
>
  <div
    class="h-full w-full flex-1 transition-all duration-300 ease-in-out bg-primary rounded-full"
    data-max="100"
    data-state="indeterminate"
    style="transform: translateX(-100%);"
  />
</div>
`;
